File: src/main/Java/boj_step16/boj12789.java
import java.io.*;
import java.util.Arrays;
import java.util.Stack;
import java.util.stream.Collectors;

public class boj12789 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine());
        int count = 1;
        Stack<Integer> line = Arrays.stream(reader.readLine().split(" ")).map(Integer::parseInt).collect(Collectors.toCollection(Stack::new));
        Stack<Integer> spare = new Stack<>();
        while (true) {
            if (line.isEmpty() && spare.isEmpty()) {
                writer.append("Nice");
                break;
            }
            if (spare.isEmpty()) {
                if (line.firstElement() != count) spare.push(line.firstElement());
                else count++;
                line.removeElementAt(0);
            } else if(line.isEmpty()) {
                if (spare.lastElement() != count) {
                    writer.append("Sad");
                    break;
                }
                spare.pop();
                count++;
            } else {
                if (line.firstElement() != count && spare.lastElement() != count) {
                    spare.add(line.firstElement());
                    line.removeElementAt(0);
                } else if (line.firstElement() == count) {
                    line.removeElementAt(0);
                    count++;
                } else if (spare.lastElement() == count) {
                    spare.pop();
                    count++;
                }
            }
        }
        writer.close();
    }
}

File: src/main/boj_step16/boj12789.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    var count = 1
    val line = Stack<Int>()
    val spare = Stack<Int>()
    val peoples = r.readLine().split(" ").map { it.toInt() }
    peoples.forEach {
        line.add(it)
    }
    while (true) {
        if (line.isEmpty() && spare.isEmpty()) {
            w.write("Nice")
            break
        }
        if (spare.isEmpty()) {
            if (line.first() != count) spare.add(line.first()) else count++
            line.removeAt(0)
        } else if (line.isEmpty()) {
            if (spare.last() != count) {
                w.write("Sad")
                break
            }
            spare.pop()
            count++
        } else {
            if (line.first() != count && spare.last() != count) {
                spare.add(line.first())
                line.removeAt(0)
            } else if (line.first() == count) {
                line.removeAt(0)
                count++
            } else if (spare.last() == count) {
                spare.pop()
                count++
            }
        }
    }
    w.close()
}